File: tests/moca-listen-output.py
#!/usr/bin/env python3
"""Observe the listener with real curl/SSE and isolated audio players."""
import json
import os
from pathlib import Path
import queue
import re
import signal
import subprocess
import sys
import tempfile
import threading
import time
import unittest
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

ROOT = Path(__file__).resolve().parents[1]
LISTENER = Path(sys.argv.pop(1)).resolve() if len(sys.argv) > 1 and not sys.argv[1].startswith('-') else ROOT / 'bin/moca-listen'
STAMP = re.compile(r'^\[\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}[+-]\d{4}\] moca-listen: ')


class Stream:
    def __init__(self, status=200, content_type='text/event-stream'):
        self.status = status
        self.content_type = content_type
        self.chunks = queue.Queue()
        self.opened = threading.Event()


class ListenerOutput(unittest.TestCase):
    def setUp(self):
        self.tmp = tempfile.TemporaryDirectory()
        self.addCleanup(self.tmp.cleanup)
        self.directory = Path(self.tmp.name)
        self.streams = queue.Queue()
        self.bodies = []
        self.connections = []
        self.process = None
        owner = self

        class Handler(BaseHTTPRequestHandler):
            def log_message(self, *_):
                pass

            def do_GET(self):
                try:
                    stream = owner.streams.get(timeout=3)
                except queue.Empty:
                    self.send_error(503)
                    return
                owner.connections.append(stream)
                self.send_response(stream.status)
                # Deliberately mixed case, as HTTP headers are case-insensitive.
                self.send_header('cOnTeNt-TyPe', stream.content_type)
                self.end_headers()
                self.wfile.flush()
                stream.opened.set()
                try:
                    while True:
                        chunk = stream.chunks.get(timeout=5)
                        if chunk is None:
                            break
                        self.wfile.write(chunk)
                        self.wfile.flush()
                except (BrokenPipeError, ConnectionResetError, queue.Empty):
                    pass

            def do_POST(self):
                body = self.rfile.read(int(self.headers['Content-Length']))
                owner.bodies.append(body)
                if body == b'fetch failure':
                    self.send_error(500)
                    return
                audio = b'RIFF' + b'\xff' * 4 + b'WAVEfmt ' + bytes(24) + b'data' + bytes(8)
                self.send_response(200)
                self.send_header('Content-Length', str(len(audio)))
                self.end_headers()
                self.wfile.write(audio)

        self.server = ThreadingHTTPServer(('127.0.0.1', 0), Handler)
        self.server.daemon_threads = True
        self.thread = threading.Thread(target=self.server.serve_forever, daemon=True)
        self.thread.start()
        self.addCleanup(self.stop)
        self.stdout = self.directory / 'stdout'
        self.stderr = self.directory / 'stderr'
        self.events = self.directory / 'players'
        stub = f'''#!{sys.executable}
import json, os, pathlib, sys, time
args = sys.argv[1:]
if pathlib.Path(sys.argv[0]).name == 'ffplay':
    sys.stdin.buffer.read()
    # Reproduce the real ffplay status output measured with SDL dummy audio.
    if '-nostats' not in args:
        print()
        sys.stderr.write('\\x1b[2K\\r')
with open(os.environ['PLAYER_EVENTS'], 'a') as output:
    output.write(json.dumps(args) + '\\n')
if os.environ.get('PLAYER_GATE'):
    while not pathlib.Path(os.environ['PLAYER_GATE']).exists():
        time.sleep(.01)
if os.environ.get('PLAYER_FAIL'):
    sys.stderr.write('player: device unavailable\\n')
    sys.exit(int(os.environ.get('PLAYER_FAIL_STATUS', '7')))
'''
        for player in ('ffplay', 'afplay', 'pw-play', 'paplay', 'aplay', 'powershell.exe'):
            path = self.directory / player
            path.write_text(stub)
            path.chmod(0o755)
        wslpath = self.directory / 'wslpath'
        wslpath.write_text('#!/bin/sh\nprintf "C:/Temp/moca.wav\\n"\n')
        wslpath.chmod(0o755)

    def stop(self):
        if self.process is not None:
            try:
                os.killpg(self.process.pid, signal.SIGTERM)
            except ProcessLookupError:
                pass
            self.process.wait(timeout=3)
        for stream in self.connections:
            stream.chunks.put(None)
        self.server.shutdown()
        self.server.server_close()
        self.thread.join(timeout=3)

    def stream(self, **kwargs):
        stream = Stream(**kwargs)
        self.streams.put(stream)
        return stream

    def start(self, player='ffplay', **extra):
        env = dict(os.environ, PATH=f'{self.directory}:' + os.environ['PATH'],
                   MOCA_URL=f'http://127.0.0.1:{self.server.server_port}',
                   MOCA_PLAYER=player, MOCA_RETRY_DELAY='0.05', MOCA_VOLUME='30',
                   PLAYER_EVENTS=str(self.events), TZ='JST-9', **extra)
        with self.stdout.open('wb') as out, self.stderr.open('wb') as err:
            self.process = subprocess.Popen([str(LISTENER)], env=env,
                                            stdout=out, stderr=err, start_new_session=True)

    def read(self, path):
        return path.read_text() if path.exists() else ''

    def wait(self, predicate):
        deadline = time.monotonic() + 4
        while time.monotonic() < deadline:
            if predicate():
                return
            time.sleep(.01)
        self.fail(f'timed out; stdout={self.read(self.stdout)!r}; stderr={self.read(self.stderr)!r}')

    def count(self, message, path=None):
        return self.read(path or self.stdout).count(message)

    def assert_clean_states(self):
        for path in (self.stdout, self.stderr):
            text = self.read(path)
            self.assertNotIn('\x1b', text)
            self.assertNotIn('\r', text)
            for line in text.splitlines():
                self.assertTrue(line.strip(), repr(text))
                if 'moca-listen:' in line:
                    self.assertRegex(line, STAMP)
                    self.assertIn('+0900]', line)

    def test_silent_connection_and_reconnection(self):
        first = self.stream()
        second = self.stream()
        self.start()
        self.wait(lambda: self.count('接続しました') == 1)
        self.assertTrue(first.opened.is_set())
        self.assertEqual(self.bodies, [])
        for _ in range(20):
            first.chunks.put(b': keepalive\r\n\r\n')
        first.chunks.put(None)
        self.wait(lambda: self.count('再接続しました') == 1)
        self.assertTrue(second.opened.is_set())
        self.assertEqual(self.count('接続しました'), 2)
        self.assertNotIn('接続しました', self.read(self.stderr))
        self.assert_clean_states()

    def test_http_failure_then_recovery(self):
        bad = self.stream(status=503)
        bad.chunks.put(None)
        good = self.stream()
        self.start()
        self.wait(lambda: self.count('再接続しました') == 1)
        self.assertTrue(good.opened.is_set())
        self.assertEqual(self.count('接続しました'), 1)
        self.assertIn('接続できません', self.read(self.stderr))
        self.assertIn('503', self.read(self.stderr))
        self.assert_clean_states()

    def test_non_sse_response_is_not_connected(self):
        bad = self.stream(content_type='text/html')
        bad.chunks.put(b'data: must not play\n\n')
        bad.chunks.put(None)
        self.stream()
        self.start()
        self.wait(lambda: self.count('再接続しました') == 1)
        self.assertEqual(self.count('接続しました'), 1)
        self.assertEqual(self.bodies, [])
        self.assertIn('SSE', self.read(self.stderr))
        self.assert_clean_states()

    def test_notifications_are_ordered_and_success_follows_player_exit(self):
        stream = self.stream()
        gate = self.directory / 'gate'
        self.start(PLAYER_GATE=str(gate))
        stream.chunks.put(b': ping\n\ndata: private first\ndata: second line\n\n: ping\n\ndata: next\n\ndata: last\n\n')
        self.wait(lambda: self.events.exists())
        self.assertEqual(self.count('通知を受信しました'), 1)
        self.assertEqual(self.count('通知を再生しました'), 0)
        self.assertEqual(self.bodies, [b'private first\nsecond line'])
        gate.touch()
        self.wait(lambda: self.count('通知を再生しました') == 3)
        self.assertEqual(self.bodies, [b'private first\nsecond line', b'next', b'last'])
        messages = [line.split('moca-listen: ')[1] for line in self.read(self.stdout).splitlines()
                    if '通知' in line]
        self.assertEqual(messages, ['通知を受信しました', '通知を再生しました'] * 3)
        self.assertNotIn('private first', self.read(self.stdout) + self.read(self.stderr))
        for args in map(json.loads, self.read(self.events).splitlines()):
            self.assertEqual(args[args.index('-volume') + 1], '30')
        self.assert_clean_states()

    def test_player_failure_keeps_diagnostics_and_continues(self):
        stream = self.stream()
        self.start(PLAYER_FAIL='1')
        stream.chunks.put(b'data: one\n\ndata: two\n\n')
        self.wait(lambda: self.count('通知の再生に失敗しました', self.stderr) == 2)
        self.assertEqual(self.count('通知を受信しました'), 2)
        self.assertEqual(self.count('通知を再生しました'), 0)
        self.assertEqual(self.count('player: device unavailable', self.stderr), 2)
        self.assert_clean_states()

    def test_ffplay_error_with_zero_exit_is_not_success(self):
        stream = self.stream()
        self.start(PLAYER_FAIL='1', PLAYER_FAIL_STATUS='0')
        stream.chunks.put(b'data: broken audio\n\n')
        self.wait(lambda: self.count('通知の再生に失敗しました', self.stderr) == 1)
        self.assertEqual(self.count('通知を再生しました'), 0)
        self.assertIn('player: device unavailable', self.read(self.stderr))
        self.assert_clean_states()

    def test_native_players_and_fetch_failure(self):
        for player in ('afplay', 'pw-play', 'paplay', 'aplay', 'windows-soundplayer'):
            with self.subTest(player=player):
                stream = self.stream()
                self.start(player)
                stream.chunks.put(b'data: native\n\ndata: fetch failure\n\n')
                self.wait(lambda: self.count('通知の再生に失敗しました', self.stderr) == 1)
                self.assertEqual(self.count('通知を受信しました'), 2)
                self.assertEqual(self.count('通知を再生しました'), 1)
                self.assertIn('500', self.read(self.stderr))
                self.assert_clean_states()
                os.killpg(self.process.pid, signal.SIGTERM)
                self.assertEqual(self.process.wait(timeout=3), -signal.SIGTERM)
                stream.chunks.put(None)
                self.process = None

    def test_sigint_stops_playing_listener(self):
        stream = self.stream()
        gate = self.directory / 'gate'
        self.start(PLAYER_GATE=str(gate), TMPDIR=str(self.directory))
        stream.chunks.put(b'data: playing\n\n')
        self.wait(lambda: self.events.exists())
        os.killpg(self.process.pid, signal.SIGINT)
        self.assertEqual(self.process.wait(timeout=3), -signal.SIGINT)
        self.assertEqual(self.count('再接続します'), 0)
        self.wait(lambda: not list(self.directory.glob('moca-listen.*')))

    def test_sigint_stops_idle_listener(self):
        self.stream()
        self.start()
        self.wait(lambda: self.count('接続しました') == 1)
        os.killpg(self.process.pid, signal.SIGINT)
        self.process.wait(timeout=3)
        self.assertNotEqual(self.process.returncode, 0)


if __name__ == '__main__':
    unittest.main()

File: tests/moca_listen_output.rs
//! Observe the shipped shell listener with real curl/SSE and isolated players.
#![cfg(unix)]
use std::{
    collections::VecDeque,
    fs::{self, File},
    io::{BufRead, BufReader, Read, Write},
    net::{TcpListener, TcpStream},
    os::unix::{
        fs::PermissionsExt,
        process::{CommandExt, ExitStatusExt},
    },
    path::{Path, PathBuf},
    process::{Child, Command, ExitStatus, Stdio},
    sync::{
        atomic::{AtomicBool, Ordering},
        mpsc, Arc, Mutex,
    },
    thread::{self, JoinHandle},
    time::{Duration, Instant},
};

struct Stream {
    chunks: mpsc::Sender<Option<Vec<u8>>>,
    opened: Arc<AtomicBool>,
}
impl Stream {
    fn send(&self, text: &str) {
        self.chunks.send(Some(text.as_bytes().to_vec())).unwrap();
    }
    fn close(&self) {
        let _ = self.chunks.send(None);
    }
    fn opened(&self) -> bool {
        self.opened.load(Ordering::SeqCst)
    }
}
struct Reply {
    status: u16,
    content_type: &'static str,
    chunks: mpsc::Receiver<Option<Vec<u8>>>,
    opened: Arc<AtomicBool>,
}
#[derive(Default)]
struct Server {
    streams: Mutex<VecDeque<Reply>>,
    bodies: Mutex<Vec<String>>,
    stopped: AtomicBool,
}
impl Server {
    fn serve(&self, mut socket: TcpStream) -> std::io::Result<()> {
        socket.set_read_timeout(Some(Duration::from_secs(2)))?;
        socket.set_write_timeout(Some(Duration::from_secs(2)))?;
        let mut input = BufReader::new(socket.try_clone()?);
        let mut first = String::new();
        input.read_line(&mut first)?;
        let mut length = 0;
        loop {
            let mut header = String::new();
            if input.read_line(&mut header)? == 0 || header == "\r\n" {
                break;
            }
            if let Some((key, value)) = header.split_once(':') {
                if key.eq_ignore_ascii_case("content-length") {
                    length = value.trim().parse::<usize>().unwrap();
                }
            }
        }
        if first.starts_with("GET ") {
            let reply = self.streams.lock().unwrap().pop_front();
            let Some(reply) = reply else {
                return socket.write_all(b"HTTP/1.0 503 Unavailable\r\nContent-Length: 0\r\n\r\n");
            };
            // Keep mixed-case headers: detecting a non-SSE response must be case-insensitive.
            write!(
                socket,
                "HTTP/1.0 {} Fixture\r\ncOnTeNt-TyPe: {}\r\n\r\n",
                reply.status, reply.content_type
            )?;
            socket.flush()?;
            reply.opened.store(true, Ordering::SeqCst);
            while !self.stopped.load(Ordering::SeqCst) {
                match reply.chunks.recv_timeout(Duration::from_millis(50)) {
                    Ok(Some(bytes)) => {
                        socket.write_all(&bytes)?;
                        socket.flush()?;
                    }
                    Ok(None) | Err(mpsc::RecvTimeoutError::Disconnected) => break,
                    Err(mpsc::RecvTimeoutError::Timeout) => {}
                }
            }
        } else {
            assert!(first.starts_with("POST "), "{first}");
            let mut body = vec![0; length];
            input.read_exact(&mut body)?;
            self.bodies
                .lock()
                .unwrap()
                .push(String::from_utf8(body.clone()).unwrap());
            if body == b"fetch failure" {
                socket.write_all(b"HTTP/1.0 500 Fixture\r\nContent-Length: 0\r\n\r\n")?;
            } else {
                let audio = [
                    b"RIFF".as_slice(),
                    &[255; 4],
                    b"WAVEfmt ",
                    &[0; 24],
                    b"data",
                    &[0; 8],
                ]
                .concat();
                write!(
                    socket,
                    "HTTP/1.0 200 OK\r\nContent-Length: {}\r\n\r\n",
                    audio.len()
                )?;
                socket.write_all(&audio)?;
            }
        }
        Ok(())
    }
}

struct Listener {
    directory: tempfile::TempDir,
    server: Arc<Server>,
    server_thread: Option<JoinHandle<()>>,
    port: u16,
    child: Option<Child>,
}
impl Listener {
    fn new() -> Self {
        let directory = tempfile::tempdir().unwrap();
        for player in [
            "ffplay",
            "afplay",
            "pw-play",
            "paplay",
            "aplay",
            "powershell.exe",
        ] {
            executable(
                &directory.path().join(player),
                include_str!("support/player.sh"),
            );
        }
        executable(
            &directory.path().join("wslpath"),
            "#!/bin/sh\nprintf 'C:/Temp/moca.wav\\n'\n",
        );
        let server = Arc::new(Server::default());
        let listener = TcpListener::bind(("127.0.0.1", 0)).unwrap();
        listener.set_nonblocking(true).unwrap();
        let port = listener.local_addr().unwrap().port();
        let state = server.clone();
        let server_thread = thread::spawn(move || {
            let mut connections = Vec::new();
            while !state.stopped.load(Ordering::SeqCst) {
                match listener.accept() {
                    Ok((socket, _)) => {
                        let state = state.clone();
                        connections.push(thread::spawn(move || {
                            let _ = state.serve(socket);
                        }));
                    }
                    Err(e) if e.kind() == std::io::ErrorKind::WouldBlock => {
                        thread::sleep(Duration::from_millis(5))
                    }
                    Err(e) => panic!("{e}"),
                }
            }
            for connection in connections {
                connection.join().unwrap();
            }
        });
        Self {
            directory,
            server,
            server_thread: Some(server_thread),
            port,
            child: None,
        }
    }
    fn path(&self, file: &str) -> PathBuf {
        self.directory.path().join(file)
    }
    fn stream(&self, status: u16, content_type: &'static str) -> Stream {
        let (chunks, receiver) = mpsc::channel();
        let opened = Arc::new(AtomicBool::new(false));
        self.server.streams.lock().unwrap().push_back(Reply {
            status,
            content_type,
            chunks: receiver,
            opened: opened.clone(),
        });
        Stream { chunks, opened }
    }
    fn sse(&self) -> Stream {
        self.stream(200, "text/event-stream")
    }
    fn start(&mut self, player: &str, extra: &[(&str, &str)]) {
        assert!(self.child.is_none());
        let listener = std::env::var_os("MOCA_LISTENER_UNDER_TEST")
            .map(PathBuf::from)
            .unwrap_or_else(|| Path::new(env!("CARGO_MANIFEST_DIR")).join("bin/moca-listen"));
        let path = std::env::join_paths(
            std::iter::once(self.directory.path().to_path_buf())
                .chain(std::env::split_paths(&std::env::var_os("PATH").unwrap())),
        )
        .unwrap();
        self.child = Some(
            Command::new(listener.canonicalize().unwrap())
                .process_group(0)
                .env("PATH", path)
                .env("MOCA_URL", format!("http://127.0.0.1:{}", self.port))
                .env("MOCA_PLAYER", player)
                .env("MOCA_RETRY_DELAY", "0.05")
                .env("MOCA_VOLUME", "30")
                .env("PLAYER_EVENTS", self.path("players"))
                .env("TMPDIR", self.directory.path())
                .env("TZ", "JST-9")
                .env_remove("PLAYER_GATE")
                .env_remove("PLAYER_FAIL")
                .env_remove("PLAYER_FAIL_STATUS")
                .envs(extra.iter().copied())
                .stdout(File::create(self.path("stdout")).unwrap())
                .stderr(File::create(self.path("stderr")).unwrap())
                .spawn()
                .unwrap(),
        );
    }
    fn read(&self, file: &str) -> String {
        fs::read_to_string(self.path(file)).unwrap_or_default()
    }
    fn count(&self, text: &str, file: &str) -> usize {
        self.read(file).matches(text).count()
    }
    fn bodies(&self) -> Vec<String> {
        self.server.bodies.lock().unwrap().clone()
    }
    fn wait(&self, ready: impl Fn() -> bool) {
        let end = Instant::now() + Duration::from_secs(4);
        while !ready() {
            assert!(
                Instant::now() < end,
                "timeout; stdout={:?}; stderr={:?}",
                self.read("stdout"),
                self.read("stderr")
            );
            thread::sleep(Duration::from_millis(10));
        }
    }
    fn clean_states(&self) {
        for file in ["stdout", "stderr"] {
            let output = self.read(file);
            assert!(!output.contains(['\x1b', '\r']), "{output:?}");
            for line in output.lines() {
                assert!(!line.trim().is_empty(), "{output:?}");
                if line.contains("moca-listen:") {
                    let (stamp, _) = line
                        .strip_prefix('[')
                        .unwrap()
                        .split_once("] moca-listen: ")
                        .unwrap();
                    assert!(stamp.ends_with("+0900"), "{line}");
                    let parsed =
                        chrono::DateTime::parse_from_str(stamp, "%Y-%m-%d %H:%M:%S%z").unwrap();
                    assert_eq!(parsed.format("%Y-%m-%d %H:%M:%S%z").to_string(), stamp);
                }
            }
        }
    }
    fn signal(&self, signal: &str) {
        signal_group(self.child.as_ref().unwrap().id(), signal);
    }
    fn wait_exit(&mut self) -> ExitStatus {
        let deadline = Instant::now() + Duration::from_secs(3);
        loop {
            if let Some(status) = self.child.as_mut().unwrap().try_wait().unwrap() {
                return status;
            }
            assert!(Instant::now() < deadline, "listener failed to stop");
            thread::sleep(Duration::from_millis(10));
        }
    }
    fn stop(&mut self) {
        if let Some(mut child) = self.child.take() {
            signal_group(child.id(), "TERM");
            let deadline = Instant::now() + Duration::from_secs(3);
            while child.try_wait().unwrap().is_none() && Instant::now() < deadline {
                thread::sleep(Duration::from_millis(10));
            }
            if child.try_wait().unwrap().is_none() {
                signal_group(child.id(), "KILL");
            }
            let _ = child.wait();
        }
    }
}
impl Drop for Listener {
    fn drop(&mut self) {
        self.stop();
        self.server.stopped.store(true, Ordering::SeqCst);
        if let Some(thread) = self.server_thread.take() {
            let _ = thread.join();
        }
    }
}
fn signal_group(pid: u32, signal: &str) {
    let _ = Command::new("kill")
        .args(["-s", signal, "--", &format!("-{pid}")])
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .status();
}
fn executable(path: &Path, text: &str) {
    fs::write(path, text).unwrap();
    fs::set_permissions(path, fs::Permissions::from_mode(0o755)).unwrap();
}

#[test]
fn silent_connection_and_reconnection() {
    let mut l = Listener::new();
    let first = l.sse();
    let second = l.sse();
    l.start("ffplay", &[]);
    l.wait(|| l.count("接続しました", "stdout") == 1);
    assert!(first.opened());
    assert!(l.bodies().is_empty());
    for _ in 0..20 {
        first.send(": keepalive\r\n\r\n");
    }
    first.close();
    l.wait(|| l.count("再接続しました", "stdout") == 1);
    assert!(second.opened());
    assert_eq!(l.count("接続しました", "stdout"), 2);
    assert!(!l.read("stderr").contains("接続しました"));
    l.clean_states();
}
#[test]
fn http_failure_then_recovery() {
    let mut l = Listener::new();
    let bad = l.stream(503, "text/event-stream");
    bad.close();
    let good = l.sse();
    l.start("ffplay", &[]);
    l.wait(|| l.count("再接続しました", "stdout") == 1);
    assert!(good.opened());
    assert_eq!(l.count("接続しました", "stdout"), 1);
    assert!(l.read("stderr").contains("接続できません"));
    assert!(l.read("stderr").contains("503"));
    l.clean_states();
}
#[test]
fn non_sse_response_is_not_connected() {
    let mut l = Listener::new();
    let bad = l.stream(200, "text/html");
    bad.send("data: must not play\n\n");
    bad.close();
    let _good = l.sse();
    l.start("ffplay", &[]);
    l.wait(|| l.count("再接続しました", "stdout") == 1);
    assert_eq!(l.count("接続しました", "stdout"), 1);
    assert!(l.bodies().is_empty());
    assert!(l.read("stderr").contains("SSE"));
    l.clean_states();
}
#[test]
fn notifications_are_ordered_and_success_follows_player_exit() {
    let mut l = Listener::new();
    let stream = l.sse();
    let gate = l.path("gate");
    l.start("ffplay", &[("PLAYER_GATE", gate.to_str().unwrap())]);
    stream.send(": ping\n\ndata: private first\ndata: second line\n\n: ping\n\ndata: next\n\ndata: last\n\n");
    l.wait(|| l.path("players").exists());
    assert_eq!(l.count("通知を受信しました", "stdout"), 1);
    assert_eq!(l.count("通知を再生しました", "stdout"), 0);
    assert_eq!(l.bodies(), ["private first\nsecond line"]);
    fs::write(gate, "").unwrap();
    l.wait(|| l.count("通知を再生しました", "stdout") == 3);
    assert_eq!(l.bodies(), ["private first\nsecond line", "next", "last"]);
    let output = l.read("stdout");
    let messages: Vec<_> = output
        .lines()
        .filter(|line| line.contains("通知"))
        .map(|line| line.split_once("moca-listen: ").unwrap().1)
        .collect();
    assert_eq!(
        messages,
        ["通知を受信しました", "通知を再生しました"].repeat(3)
    );
    assert!(!(output + &l.read("stderr")).contains("private first"));
    let records = l.read("players");
    for record in records.split("\0\0").filter(|s| !s.is_empty()) {
        let args: Vec<_> = record.split('\0').collect();
        assert_eq!(
            args[args.iter().position(|a| *a == "-volume").unwrap() + 1],
            "30"
        );
    }
    l.clean_states();
}
#[test]
fn player_failure_keeps_diagnostics_and_continues() {
    let mut l = Listener::new();
    let stream = l.sse();
    l.start("ffplay", &[("PLAYER_FAIL", "1")]);
    stream.send("data: one\n\ndata: two\n\n");
    l.wait(|| l.count("通知の再生に失敗しました", "stderr") == 2);
    assert_eq!(l.count("通知を受信しました", "stdout"), 2);
    assert_eq!(l.count("通知を再生しました", "stdout"), 0);
    assert_eq!(l.count("player: device unavailable", "stderr"), 2);
    l.clean_states();
}
#[test]
fn ffplay_error_with_zero_exit_is_not_success() {
    let mut l = Listener::new();
    let stream = l.sse();
    l.start(
        "ffplay",
        &[("PLAYER_FAIL", "1"), ("PLAYER_FAIL_STATUS", "0")],
    );
    stream.send("data: broken audio\n\n");
    l.wait(|| l.count("通知の再生に失敗しました", "stderr") == 1);
    assert_eq!(l.count("通知を再生しました", "stdout"), 0);
    assert!(l.read("stderr").contains("player: device unavailable"));
    l.clean_states();
}
#[test]
fn native_players_and_fetch_failure() {
    for player in [
        "afplay",
        "pw-play",
        "paplay",
        "aplay",
        "windows-soundplayer",
    ] {
        let mut l = Listener::new();
        let stream = l.sse();
        l.start(player, &[]);
        stream.send("data: native\n\ndata: fetch failure\n\n");
        l.wait(|| l.count("通知の再生に失敗しました", "stderr") == 1);
        assert_eq!(l.count("通知を受信しました", "stdout"), 2);
        assert_eq!(l.count("通知を再生しました", "stdout"), 1);
        assert!(l.read("stderr").contains("500"));
        l.clean_states();
        l.signal("TERM");
        assert_eq!(l.wait_exit().signal(), Some(15));
        stream.close();
    }
}
#[test]
fn sigint_stops_playing_listener() {
    let mut l = Listener::new();
    let stream = l.sse();
    let gate = l.path("gate");
    let tmp = l.directory.path().to_path_buf();
    l.start(
        "ffplay",
        &[
            ("PLAYER_GATE", gate.to_str().unwrap()),
            ("TMPDIR", tmp.to_str().unwrap()),
        ],
    );
    stream.send("data: playing\n\n");
    l.wait(|| l.path("players").exists());
    l.signal("INT");
    assert_eq!(l.wait_exit().signal(), Some(2));
    assert_eq!(l.count("再接続します", "stdout"), 0);
    l.wait(|| {
        fs::read_dir(&tmp).unwrap().all(|e| {
            !e.unwrap()
                .file_name()
                .to_string_lossy()
                .starts_with("moca-listen.")
        })
    });
}
#[test]
fn sigint_stops_idle_listener() {
    let mut l = Listener::new();
    let _stream = l.sse();
    l.start("ffplay", &[]);
    l.wait(|| l.count("接続しました", "stdout") == 1);
    l.signal("INT");
    assert!(!l.wait_exit().success());
}
